### PR TITLE
#506: Add capability to create and export a dataset containing shared blocks

### LINKED ISSUE #506: Add capability to create and export a dataset containing shared blocks
Conceptually, this would be allow for the creation of JSON files representing such diagrams:

Screen Shot 2024-05-03 at 5 58 07 PM

### PR DESCRIPTION
**Pull request by tlamonthezie**
Fixes #506 

This PR provides an utility script to generate a dataset from a phase specification (described by a dictionary representing a  diagram, including tasks, shared blocks, communications, communicators, and ranks)

It includes
- A `PhaseSpecification` typed dictionary to represent the diagram in some program-readable format (dictionary).
- The implementation of a `populate_from_specification` method (Phase class)
- The addition of the `lbaf-vt-data-files-maker` command to the the package + unit tests for valid/invalid configurations
- Extra minor changes: CI actions version upgrades

A phase specification can be described using either JSON or YAML format
_(in both cases please note that python sets are represented in a specific way because it is not native to JSON and YAML formats)_

<details>

<summary>Specification sample (JSON Format)</summary>

```json
{
    "communications": [
        {
            "from": 0,
            "size": 10000.0,
            "to": 2
        },
        {
            "from": 1,
            "size": 15000.0,
            "to": 2
        },
        {
            "from": 2,
            "size": 20000.0,
            "to": 1
        },
        {
            "from": 0,
            "size": 25000.0,
            "to": 1
        }
    ],
    "ranks": {
        "0": {
            "tasks": [
                0,
                1
            ]
        },
        "1": {
            "tasks": [
                2
            ]
        }
    },
    "shared_blocks": [
        {
            "size": 10000.0,
            "tasks": [
                0,
                1
            ],
            "home_rank": 0
        },
        {
            "size": 15000.0,
            "tasks": [
                2
            ],
            "home_rank": 1
        }
    ],
    "tasks": [
        2.0,
        3.5,
        5.0
    ]
}
```

</details>

<details>

<summary>Specification sample (YAML Format)</summary>

```yaml
communications:
  - {from: 0, size: 10000.0, to: 2}
  - {from: 1, size: 15000.0, to: 2}
  - {from: 2, size: 20000.0, to: 1}
  - {from: 0, size: 25000.0, to: 1}
ranks:
  0:
    tasks: [0, 1]
  1:
    tasks: [2]
shared_blocks:
  - size: 10000.0
    tasks: [0, 1]
    home_rank: 0
  - size: 15000.0
    tasks: [2]
    home_rank: 1
tasks: [2.0, 3.5, 5.0]

```

</details>

Once the specification file run by setting at minimum the specification file path and the writer output data stem:
`lbaf-vt-data-files-maker --spec-file=/home/john/data-maker/dataset1.json --data-stem=/home/john/data-maker/dataset1`

An interactive mode is also available as an option by calling `lbaf-vt-data-files-maker --interactive` providing more tools to create, update, delete, print a specification.

Run `lbaf-vt-data-files-maker --help` to get all available options

### CODE REVIEW

**Comment by ppebay** [comment]
@tlamonthezie the proposed schema makes sense to me but I have some questions:
* what is the role of the `null` values in the YAML descriptions of the communication edges? 
* use consistent indentations after the `-` symbols (makes it much easier to parse at least in VIM)
* regarding the use of the `!!set` data type specifiers, I am not opposed to it but I am unsure as to whether it's useful or not, because nothing would prevent a typo at this level, resulting in a singleton set (as opposed to a proper pair) and if no YAML validation is done, this would result in a runtime error in the code anyway. Thoughts on that?
* _"python sets are represented in a specific way"_: which specific way exactly?

**Comment by tlamonthezie** [comment]
> > @tlamonthezie the proposed schema makes sense to me but I have some questions:
> > 
> > * what is the role of the `null` values in the YAML descriptions of the communication edges?
> > * use consistent indentations after the `-` symbols (makes it much easier to parse at least in VIM)
> > * regarding the use of the `!!set` data type specifiers, I am not opposed to it but I am unsure as to whether it's useful or not, because nothing would prevent a typo at this level, resulting in a singleton set (as opposed to a proper pair) and if no YAML validation is done, this would result in a runtime error in the code anyway. Thoughts on that?
> > * _"python sets are represented in a specific way"_: which specific way exactly?
> 
> Hi @ppebay
> 
> Globally your questions about the format of the specification file correspond to things that in my opinion can still be made better. So I agree with that.
> 
> * Yaml null values are not looking good in my opinion and I will work to change that. In fact this is the way that pyyaml represents Python sets but I think that for a better look we could represent that sets as lists. In our case the specification defines tasks as a sets in a shared block definition because a set contains unique values.
> * Ok to use consistent indentations i understand this is some pyyaml default to override
> * The `!!set` is also generated by pyyaml and it seems to be part of the yaml specification to represent sets ( https://yaml.org/spec/1.2.2/#10111-generic-mapping) as we can read that "Sets are represented as a Mapping where each key is associated with a null value". But as in point 1 I think it will be better to use just a list representation and convert to sets at runtime.
>   I will continue to work on a better formatting
> * The `specific way` I was talking about is just the flags needed in json or in yaml to say that some data represents a `set` (in yaml for example this is the `!!set` mark)
> 
> Regarding the new `populate_from_specification` method in the `Phase` class, did you @ppebay already check the logic implementation ?

Hi @ppebay I have just updated the PR comment with updated format. Yaml is more concise and simple and use now lists to represents the sets. Could you please take a look again ?

**Comment by ppebay** [comment]
Thanks @tlamonthezie, for the detailed answers; I agree about the replacement of sets by lists (and in any event I believe the natural way to handle consistency is via a validator like is done elsewhere throughout the code).

**Comment by ppebay** [comment]
@tlamonthezie there is still something I do not understand about the specification: for instance in YAML, 
```
ranks:
  0:
    communications: [0, 3]
    tasks: [0, 1]
  1:
    communications: [1, 2]
    tasks: [2]
```
when it comes to the **communications**. 

Communications do not belong to ranks, but are attached to objects. So the following is correct:
```
communications:
  - {from: 0, size: 10000.0, to: 2}
  - {from: 1, size: 15000.0, to: 2}
  - {from: 2, size: 20000.0, to: 1}
  - {from: 0, size: 25000.0, to: 1}
 ```

Therefore in this context what is the meaning of assigning communications `[0, 3]` to rank 0, and `[1, 2]` to rank 1?

**Comment by tlamonthezie** [comment]
> @tlamonthezie there is still something I do not understand about the specification: for instance in YAML,
> 
> ```
> ranks:
>   0:
>     communications: [0, 3]
>     tasks: [0, 1]
>   1:
>     communications: [1, 2]
>     tasks: [2]
> ```
> 
> when it comes to the **communications**.
> 
> Communications do not belong to ranks, but are attached to objects. So the following is correct:
> 
> ```
> communications:
>   - {from: 0, size: 10000.0, to: 2}
>   - {from: 1, size: 15000.0, to: 2}
>   - {from: 2, size: 20000.0, to: 1}
>   - {from: 0, size: 25000.0, to: 1}
> ```
> 
> Therefore in this context what is the meaning of assigning communications `[0, 3]` to rank 0, and `[1, 2]` to rank 1?
> 
> I think, however, that I understand where the confusion may be stemming from, in the following diagram:
> 
> 327889642-3c285ebd-9b3e-4b98-8119-f4914a783eb6
> 
> **the placement of the communication "stickers" (labels), inside the per-rank boxes, is only a matter of visual/graphical convenience, and has no rank membership meaning.**
> 
> Those labels could have been drawn elsewhere, but were merely placed as such to avoid difficult-to-read stickers laying on top of rank box edges..

Ok @ppebay I understand your point. The VTDataWriter needs communications per rank so it was done as in the schema.
But yes we can avoid defining it like that if we determine how to index communications per rank for the writer. Should it be the rank owning the 'from' task, 'to' task, both ? If source and target tasks are in à different rank should the communication appear in both ranks (In the output should the communication appears in 2 files) ?

**Comment by ppebay** [comment]
@tlamonthezie thanks, so let's reformulate this one last time to make sure we fully understand each other; those entries:
```
    communications: [0, 3]
...
    communications: [1, 2]
```
will be **entirely removed** from the YAML descriptor, as well as their JSON counterparts:
```
"communications": [
                0,
                3
            ],
...
"communications": [
                1,
                2
            ],
```
agreed?

On this topic, it is also worth noticing one assumption regarding the data files that are coming out of vt, and that are processed by the data reader, is that communication edges are stored either in the sender or the receiving rank file, and not any other rank file. This is just a practicality in terms of I/O, but it might have be created the impression that communication edges themselves belong to ranks (which is not the case, as is worth repeating).

On another topic, when trying to execute the configuration file I added to build the diagram guiding #506 (`ccm-example.yaml`), I am getting the following r/t error:
```
python LBAF_app.py  -c ccm-example.yaml
Traceback (most recent call last):
  File "/Users/pppebay/Documents/Git/LB-analysis-framework/src/lbaf/Applications/LBAF_app.py", line 19, in <module>
    import lbaf.IO.lbsStatistics as lbstats
  File "/Users/pppebay/Documents/Git/LB-analysis-framework/src/lbaf/__init__.py", line 24, in <module>
    from lbaf.Utils.lbsJSONDatasetMaker import JSONDatasetMaker
  File "/Users/pppebay/Documents/Git/LB-analysis-framework/src/lbaf/Utils/lbsJSONDatasetMaker.py", line 327
    comm["size"] = self.__prompt.prompt("Communication size ?", required=True, value_type=float),
                 ^
SyntaxError: invalid syntax
```
Can you please look into this?

**Comment by tlamonthezie** [comment]
> python LBAF_app.py  -c ccm-example.yaml

The syntax error is due to the fact that the commit was in WIP state so a development was in progress. The error should disappear now.

**Comment by ppebay** [comment]
Thanks but regarding this: _ccm data is missing shared blocks information_, I want to add that blocks are of course optional. We may have cases where no shared memory blocks are present, and these should be supported as well.

> > python LBAF_app.py  -c ccm-example.yaml
> 
> Hi @ppebay the syntax error was due to the fact that the commit was in WIP state so a development was in progress. The error should disappear now. But there is another error appearing now when running the LBAF app and apparently due to the fact that ccm data is missing shared blocks information.

**Comment by tlamonthezie** [comment]
> @tlamonthezie there is still something I do not understand about the specification: for instance in YAML,
> 
> ```
> ranks:
>   0:
>     communications: [0, 3]
>     tasks: [0, 1]
>   1:
>     communications: [1, 2]
>     tasks: [2]
> ```
> 
> when it comes to the **communications**.
> 
> Communications do not belong to ranks, but are attached to objects. So the following is correct:
> 
> ```
> communications:
>   - {from: 0, size: 10000.0, to: 2}
>   - {from: 1, size: 15000.0, to: 2}
>   - {from: 2, size: 20000.0, to: 1}
>   - {from: 0, size: 25000.0, to: 1}
> ```
> 
> Therefore in this context what is the meaning of assigning communications `[0, 3]` to rank 0, and `[1, 2]` to rank 1?
> 
> I think, however, that I understand where the confusion may be stemming from, in the following diagram:
> 
> 327889642-3c285ebd-9b3e-4b98-8119-f4914a783eb6
> 
> **the placement of the communication "stickers" (labels), inside the per-rank boxes, is only a matter of visual/graphical convenience, and has no rank membership meaning.**
> 
> Those labels could have been drawn elsewhere, but were merely placed as such to avoid difficult-to-read stickers laying on top of rank box edges..

Hi @ppebay The specification is now updated and communications are not owned by ranks any more. The specification has been updated accordingly in the primary PR comment and communications are still output in the rank data file owning the sender task.

**Comment by ppebay** [comment]
Regarding the user-defined data, two important clarification points @tlamonthezie: 
1. the `user_defined` key allows for **optional** sub-keys; therefore, no such sub-key should be expected;
2. however, when the `shared_bytes` sub-key is provided it should always come as part of a triple, along with `shared_id` and `home`.

**Comment by ppebay** [comment]
@tlamonthezie I confirm that `ccm-example.yaml` (along with the synthetic data set) correctly builds the ranks + objects + communications (but no blocks at this time):
output_file0

**N.B.:** there is an issue in the rendering of the comm edges above, which I believe is due to some incorrect interplay between VTK and vt-tv on my system (reported in [vt-tv #32](https://github.com/orgs/DARMA-tasking/projects/9/views/2?pane=issue&itemId=62165008))

**Comment by ppebay** [comment]
@tlamonthezie regarding the presence of shared blocks inside the generated rank data files:
* shared memory blocks are identified by a unique ID (i.e. 1 and 2 in the above diagram);
* if a task (=objects in LBAF) accesses a given shared block, then this block must exist in the user-defined data of that task.

Following that logic, it follows that a shared block will appear as many times as it is shared by tasks. In turn, this will result in that block possibly appearing in several rank data files (if it is accessed by tasks residing on different ranks).

**Comment by tlamonthezie** [comment]
> @tlamonthezie regarding the presence of shared blocks inside the generated rank data files:
> 
> * shared memory blocks are identified by a unique ID (i.e. 1 and 2 in the above diagram);
> * if a task (=objects in LBAF) accesses a given shared block, then this block must exist in the user-defined data of that task.
> 
> Following that logic, it follows that a shared block will appear as many times as it is shared by tasks. In turn, this will result in that block possibly appearing in several rank data files (if it is accessed by tasks residing on different ranks).

@ppebay Just a question. Because the block must appear on each tasks, in case of tasks are on different ranks what value should be put as then the "home" in user_defined data ? The first task home, the current task home (then we create a difference between user_defined data for the same shared block)?

**Comment by ppebay** [comment]
@tlamonthezie the "home" (i.e. the ID of the unique rank to which a shared block ultimately belongs -- and to which it must be returned post-distributed calculations) is a constant and must thus be specified at block creation. This is what I was alluding to earlier:
"when the shared_bytes sub-key is provided it should always come as part of a triple, along with shared_id and home."

So to make this clear: when a shared block is specified, this must be done with a **triple**: `(ID, shared_bytes, home)`. 

In terms of validation, this means that for a given block ID, `(shared_bytes, home)` are constant in the output rank data files.

> 
> @ppebay Just a question. Because the block must appear on each tasks, in case of tasks are on different ranks what value should be put as then the "home" in user_defined data ? The first task home (currently the case), or the task home (then we create a difference between user_defined data for the same shared block between the 2 tasks) ?

**Comment by ppebay** [comment]
@tlamonthezie is the program still called `lbaf-vt-data-files-maker`?

Isn't it rather `test_json_data_files_maker.py`?

File: src/lbaf/Utils/lbsJSONDataFilesMaker.py [hunk 1]
```diff
@@ -0,0 +1,569 @@
+"""
+Utility to create and export a data set supporting shared blocks by using a specification file.
+
+To call this script either call script with
+- `lbaf-vt-data-files-maker` if lbaf package is installed or
+- `python src/lbaf/Utils/lbsJSONDatasetMaker.py`
```
Comment (ppebay) [inline src/lbaf/Utils/lbsJSONDataFilesMaker.py:6]: The name seems to be `lbsJSONDataFilesMaker.py`

**Review by ppebay — COMMENTED** [review]
I am getting the following execution error 
```
~/Documents/Git/LB-analysis-framework$ python src/lbaf/Utils/lbsJSONDataFilesMaker.py --help 
Traceback (most recent call last):
  File "/Users/pppebay/Documents/Git/LB-analysis-framework/src/lbaf/Utils/lbsJSONDataFilesMaker.py", line 33, in <module>
    from lbaf.Model.lbsPhase import Phase
ModuleNotFoundError: No module named 'lbaf'
```
please let me know what I am doing incorrectly

File: src/lbaf/Utils/lbsJSONDataFilesMaker.py (hunk 1 above)
Comment (tlamonthezie) [reply src/lbaf/Utils/lbsJSONDataFilesMaker.py]: Oh sorry you are right ! It has been fixed !

**Comment by ppebay** [comment]
@tlamonthezie still getting the same error in the non-install mode:
```
 python src/lbaf/Utils/lbsJSONDataFilesMaker.py --spec-file=tests/unit/config/phases/phase_spec_03_valid.yaml --data-stem=tests/unit/output/JSON_data_files_maker/dataset01 
Traceback (most recent call last):
  File "/Users/pppebay/Documents/Git/LB-analysis-framework/src/lbaf/Utils/lbsJSONDataFilesMaker.py", line 33, in <module>
    from lbaf.Model.lbsPhase import Phase
ModuleNotFoundError: No module named 'lbaf'
```

**Comment by tlamonthezie** [comment]
> @tlamonthezie still getting the same error in the non-install mode:
> 
> ```
>  python src/lbaf/Utils/lbsJSONDataFilesMaker.py --spec-file=tests/unit/config/phases/phase_spec_03_valid.yaml --data-stem=tests/unit/output/JSON_data_files_maker/dataset01 
> Traceback (most recent call last):
>   File "/Users/pppebay/Documents/Git/LB-analysis-framework/src/lbaf/Utils/lbsJSONDataFilesMaker.py", line 33, in <module>
>     from lbaf.Model.lbsPhase import Phase
> ModuleNotFoundError: No module named 'lbaf'
> ```

Ok @ppebay my fault the lbaf import (case lbak package not installed) was not at the correct position and I did not have the error because my lbaf pachage was installed in edit mode on my machine. This is now fixed and it should also work now on your side also.

**Comment by ppebay** [comment]
Not sure I understand @tlamonthezie -- will you push some changes as a result?
> Ok @ppebay my fault the lbaf import (case lbak package not installed) was not at the correct position and I did not have the error because my lbaf pachage was installed in edit mode on my machine. This is now fixed and it should also work now on your side.

File: src/lbaf/Execution/lbsPhaseSpecification.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                    "tasks": list(r.get("tasks", {}))
+                }
+            )
+        }
+
+    def denormalize(self, data: dict)-> PhaseSpecification:
+        """Creates a phase specification using a definition where sets are represented as lists
+        """
+
+        return PhaseSpecification({
```
Comment (ppebay) [inline src/lbaf/Execution/lbsPhaseSpecification.py:91]: Please document this block of code with a comment (1 line) as it's not clear to me what it's doing

File: src/lbaf/Model/lbsPhase.py [hunk 3]
```diff
@@ -406,6 +407,143 @@ def populate_from_log(self, phase_id):
...
+        for rank_id, rank_specs in spec["ranks"].items():
+            task_ids = rank_specs["tasks"]
+            for task_id in task_ids:
+                task_user_defined = {}
+                time = spec["tasks"][task_id]
+
+                # Check: task must be only in 1 rank
+                if task_id in objects:
+                    raise RuntimeError(f"Task already in rank {objects[task_id].get_rank_id()}")
+
```
Comment (ppebay) [inline src/lbaf/Model/lbsPhase.py:441]: Add comment here or remove blank line

File: src/lbaf/Execution/lbsPhaseSpecification.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+    # List of shared blocks sizes as
+    # - a dict { shared_block_1_id => shared_block_1, shared_block_2_id => shared_block_2 })
+    # - a list (shared block id is the index in the list)
+    shared_blocks: Union[List[SharedBlockSpecification],Dict[int,SharedBlockSpecification]]
+
+    # List of communications volumes as
+    # - a dict { com1_id => com1, com2_id => com2 })
+    # - a list (communication id is the index in the list)
+    communications: Union[List[CommunicationSpecification],Dict[int,CommunicationSpecification]]
```
Comment (ppebay) [inline src/lbaf/Execution/lbsPhaseSpecification.py:35]: This is complicated to understand: either break down statement or explain better in comment above

**Review by ppebay — CHANGES_REQUESTED** [review]
Due to the extent of this PR, this is a second review of an iterative review process. 

Please @tlamonthezie request another review once comments here are already addressed.

Also please request reviews from @lifflander and @nlslatt

File: src/lbaf/Execution/lbsPhaseSpecification.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    """
+
+    def __normalize_member(self, data: Union[dict,list], transform: Optional[Callable] = None) -> Union[dict,list]:
+        """Normalize a member that can be represented as a dict where key is the item key or as a list
+        where id is the index in the list"""
+
+        if isinstance(data, list):
+            return [ transform(o) for o in data ] if transform is not None else data # pylint: disable=E1133 (not-an-iterable)
+        elif isinstance(data, dict):
+            return { o_id:transform(o) for o_id, o in data.items() } if transform is not None else data # pylint: disable=E1101 (no-member)
```
[the remark this answers is not in the document]
Comment (tlamonthezie) [reply src/lbaf/Execution/lbsPhaseSpecification.py:57]: I removed that it was it seems it was not needed

File: src/lbaf/Execution/lbsPhaseSpecification.py (hunk 4 above)
Comment (tlamonthezie) [reply src/lbaf/Execution/lbsPhaseSpecification.py]: I have just updated the comments to make it more simple

File: src/lbaf/Model/lbsPhase.py [hunk 6]
```diff
@@ -406,6 +407,143 @@ def populate_from_log(self, phase_id):
...
+        # Auto-generate ids for Lists by converting to dictionaries to set indices as keys for tasks, shared blocks,
+        # and communications
+        if isinstance(spec["tasks"], list):
+            spec["tasks"] = {i:spec["tasks"][i] for i in range(len(spec["tasks"]))}
+        if isinstance(spec["shared_blocks"], list):
+            spec["shared_blocks"] = {i:spec["shared_blocks"][i] for i in range(len(spec["shared_blocks"]))}
+        if isinstance(spec["communications"], list):
+            spec["communications"] = {i:spec["communications"][i] for i in range(len(spec["communications"]))}
+
+        ranks: Dict[int,Rank] = {r_id:Rank(self.__logger, r_id) for r_id in spec.get("ranks", []).keys()}
```
[the remark this answers is not in the document]
Comment (tlamonthezie) [reply src/lbaf/Model/lbsPhase.py]: Ok

File: src/lbaf/Model/lbsPhase.py (hunk 3 above)
Comment (tlamonthezie) [reply src/lbaf/Model/lbsPhase.py:446]: @ppebay Do you mean le line 441 ? (there is just one blank line there and there is a statement on the line 442.

Comment (ppebay) [reply src/lbaf/Model/lbsPhase.py:446]: yes sorry I was unclear. Block of code below is a bit obscure IMO and a comment would help

File: src/lbaf/Execution/lbsPhaseSpecification.py (hunk 2 above)
Comment (tlamonthezie) [reply src/lbaf/Execution/lbsPhaseSpecification.py:95]: @ppebay Some detailed documentation has been added

File: src/lbaf/Execution/lbsPhaseSpecification.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+            "ranks": self.__normalize_member(
+                spec.get("ranks", {}),
+                lambda r: {
+                    "tasks": list(r.get("tasks", {}))
+                }
+            )
+        }
+
+    def denormalize(self, data: dict)-> PhaseSpecification:
+        """Creates a phase specification using a definition where sets are represented as lists
```
[the remark this answers is not in the document]
Comment (tlamonthezie) [reply src/lbaf/Execution/lbsPhaseSpecification.py]: Done

File: src/lbaf/Model/lbsPhase.py (hunk 3 above)
Comment (tlamonthezie) [reply src/lbaf/Model/lbsPhase.py:449]: I did a change for that is more clear ? Or do you want to suggest some another text ?

File: src/lbaf/Utils/lbsJSONDataFilesMaker.py [hunk 8]
```diff
@@ -0,0 +1,572 @@
...
+                    "max_objects_per_transfer": 2,
+                    "deterministic_transfer": True
+                }
+            },
+            "logging_level": "debug",
+            "output_dir": f"{PROJECT_PATH}/output",
+            "output_file_stem": "output_file"
+        }
+
+    def create_run_config_file(self, output_path: str):
```
[the remark this answers is not in the document]
Comment (tlamonthezie) [reply src/lbaf/Utils/lbsJSONDataFilesMaker.py:202]: @ppebay In fact I think the creation of this static run configuration should be removed. This was done for development purpose.

**Review by ppebay — COMMENTED** [review]
Thanks @tlamonthezie for the changes. Before approving I would like that @nlslatt or @lifflander also get a chance to review this PR

**Comment by ppebay** [comment]
Quick request @tlamonthezie could you please prepare a brief demo of the interactive mode for next LB meeting? 10 minutes. Thanks!

**Comment by ppebay** [comment]
@tlamonthezie please fix interactive mode, once done please re-submit for review

**Comment by tlamonthezie** [comment]
> @tlamonthezie please fix interactive mode, once done please re-submit for review

Ok I will fix it.

File: src/lbaf/Model/lbsPhase.py [hunk 9]
```diff
@@ -406,6 +408,151 @@ def populate_from_log(self, phase_id):
...
+                # Check: task must share 0 or 1 block
+                if not multiple_sharing and o.get_shared_block_id() is not None and o.get_shared_block_id() != shared_id:
+                    raise RuntimeError(
+                        f"Task {o.get_id()} already shared block {o.get_shared_block_id()} and cannot share additional block {shared_id}. Only 0 or 1 allowed")
+                o.set_shared_block(b)
+
+                # Initialize object user defined data
+                o.get_user_defined()["shared_id"] = b.get_id()
+                o.get_user_defined()["shared_bytes"] = b.get_size()
+                o.get_user_defined()["home"] = b.get_home_id()
```
Comment (nlslatt) [inline src/lbaf/Model/lbsPhase.py:490]: I was remembering wrong. This should be `home_rank`, not `home`, per https://github.com/DARMA-tasking/vt/blob/bc5264ab35da2a303c4db3d290dccc00eca67c83/src/vt/vrt/collection/balance/temperedlb/temperedlb.cc#L597. This change might need to be made in other places as well.

**Review by nlslatt — CHANGES_REQUESTED** [review]
The home of a shared block needs to be `home_rank` to be compatible with https://github.com/DARMA-tasking/vt/pull/2278.

File: src/lbaf/Model/lbsPhase.py (hunk 9 above)
Comment (nlslatt) [reply src/lbaf/Model/lbsPhase.py:490]: @ppebay

File: src/lbaf/Utils/lbsJSONDataFilesMaker.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ppebay) [reply src/lbaf/Utils/lbsJSONDataFilesMaker.py]: Please do so, and this config file could be retained for CI purposes to test/exercise this generator

File: src/lbaf/Model/lbsPhase.py (hunk 9 above)
Comment (ppebay) [reply src/lbaf/Model/lbsPhase.py:490]: @tlamonthezie please fix accordingly, thanks

**Comment by ppebay** [comment]
@tlamonthezie any update as to whether this is ready for new review?

**Comment by tlamonthezie** [comment]
> @tlamonthezie the `lbaf-vt-data-files-maker` does not seem to be included in the branch; is it there?

it should be

**Comment by tlamonthezie** [comment]
> The home of a shared block needs to be `home_rank` to be compatible with [DARMA-tasking/vt#2278](https://github.com/DARMA-tasking/vt/pull/2278).

Thanks @nlslatt for that precision. This has been updated in code, unit tests and also in the examples in the PR.

**Comment by ppebay** [comment]
@tlamonthezie getting this error when trying to create a **block** in interactive mode:

Screenshot 2024-07-05 at 05 03 18

I confirm however that **task** creation/edition is working

**Comment by tlamonthezie** [comment]
> @tlamonthezie getting this error when trying to create a **block** in interactive mode:
> 
> Screenshot 2024-07-05 at 05 03 18
> I confirm however that **task** creation/edition is working

Thanks @ppebay for the info. This is fixed.

**Comment by ppebay** [comment]
New error (when editing block)
Screenshot 2024-07-05 at 05 19 38

**Comment by ppebay** [comment]
@tlamonthezie I confirm that following the latest corrections, I have been able to go through the same steps and **not** encounter the error this time.

Furthermore I was able to also interactively generate the full diagram of the CCM case, excluding communications for simplicity for now (also adding homing that is not part of the CCM example in the paper).
```
----------- BEGIN YAML -----------
communications: []
ranks:
  0:
    tasks: [0, 1]
  1:
    tasks: [2]
shared_blocks:
  - home_rank: 1
    size: 90000.0
    tasks: [0, 1]
  - home_rank: 0
    size: 100000.0
    tasks: [2]
tasks: [2.0, 3.5, 5.0]

----------- END YAML -------------
```

**Comment by tlamonthezie** [comment]
@ppebay thank you for your latest comment. It seems that interactive mode is now robust enough. Additionnaly the "remove" feature has been completed to remove an element (ranks, communication, shared block, task) in interactive mode. Review is now pending.

**Review by ppebay — APPROVED** [review]

**Comment by tlamonthezie** [comment]
> @tlamonthezie @ppebay It appears that the json files in this PR were not updated after the `home` --> `home_rank` change to the code.

@nlslatt the ccm_example json data files have been updated.
Some other examples contains shared block information but it seems that there is no home_rank in it. If some another example needs to be updated could you please indicate in some another comment ?

**Comment by ppebay** [comment]
@tlamonthezie : confirmed, : the `home` key still valid in the `entity` node (although this is inconsistent and may have to be revised in the future)

> > @tlamonthezie @ppebay It appears that the json files in this PR were not updated after the `home` --> `home_rank` change to the code.
> 
> @nlslatt yes the `ccm_example` json data files have been updated. Some other examples contains `user_defined` shared block information but it seems that there is no`home` neither `home_rank` in it. If some another example needs to be updated could you please indicate in some another comment ?
> 
> Another question, just to be sure even if I think the response is Yes: is the `home` key still valid in the `entity` node or should it also be renamed to `home_rank` as in `user_defined` node ?

File: src/lbaf/Utils/lbsJSONDataFilesMaker.py [hunk 10]
```diff
@@ -0,0 +1,811 @@
...
+To call this script either call script with
+- `lbaf-vt-data-files-maker` if lbaf package is installed or
+- `python src/lbaf/Utils/lbsJSONDataFilesMaker.py`
+
+Script usage examples:
+
+- Generate dataset from specification file
+`lbaf-vt-data-files-maker --spec-file=/home/john/data-maker/dataset1-spec.yaml --data-stem=/home/john/data-maker/dataset1`
+
+- Generate dataset from specification file and sample configuration file configured tonuse the generated data stem
```
Comment (nlslatt) [inline src/lbaf/Utils/lbsJSONDataFilesMaker.py:13]: Typo `tonuse`

File: src/lbaf/Utils/lbsJSONDataFilesMaker.py [hunk 11]
```diff
@@ -0,0 +1,811 @@
...
+
+        writer = VTDataWriter(self.__logger, None, self.__args.data_stem, writer_parameters)
+        writer.write({phase.get_id(): phase})
+        self.__prompt.print_success("Dataset has been generated.")
+
+        self.__args.data_stem = data_stem
+
+    def load_sample(self, use_explicit_keys: bool = False):
+        """Create a new sample specification as represented by diagram specified in issue #506
+        This method implementation indicates also how to cfreate a specification from Python code
```
Comment (nlslatt) [inline src/lbaf/Utils/lbsJSONDataFilesMaker.py:241]: Typo `cfreate`

File: src/lbaf/Utils/lbsJSONDataFilesMaker.py [hunk 12]
```diff
@@ -0,0 +1,811 @@
...
+
+        if choice == "*Back":
+            return "*Back"
+        if choice != f"*New {object_type_name}":
+            return int(choice)
+        else:
+            return "*New"
+
+    def make_object(self, parent: Union[list, dict], object_type_name: str, default, update: Callable):
+        """Create or update an object in a a list (id is the index) or in a dict (id is the key) for interactive input
```
Comment (nlslatt) [inline src/lbaf/Utils/lbsJSONDataFilesMaker.py:368]: Typo `a a`

**Review by nlslatt — COMMENTED** [review]
I haven't finished looking through everything yet.

File: src/lbaf/Utils/lbsJSONDataFilesMaker.py [hunk 13]
```diff
@@ -0,0 +1,811 @@
...
+                    else "yaml" if path.endswith(".yaml") or path.endswith(".yml") \
+                    else None
+                if frmt is None:
+                    self.__prompt.print_error("Specification file path must end with either .json, .yml, .yaml")
+
+            if not os.path.isdir(output_dir):
+                os.makedirs(output_dir)
+
+            with open(path, "wt", encoding="utf-8") as o_file:
+                o_file.write(PhaseSpecificationNormalizer().normalize(spec), frmt)
```
[the remark this answers is not in the document]
Comment (tlamonthezie) [reply src/lbaf/Utils/lbsJSONDataFilesMaker.py]: Yes @nlslatt you are right. Some mistake due to latest refactoring. Thanks this is fixed.

File: src/lbaf/Utils/lbsJSONDataFilesMaker.py [hunk 14]
```diff
@@ -0,0 +1,811 @@
...
+        else:
+            self.run_extra_action(action)
+
+    def run(self):
+        """Run the JSONDatasetMaker"""
+
+        # Parse command line arguments
+        self.__parse_args()
+
+        spec: PhaseSpecification = self.process_args()
```
[the remark this answers is not in the document]
Comment (tlamonthezie) [reply src/lbaf/Utils/lbsJSONDataFilesMaker.py]: This has been also fixed. Thanks @nlslatt.

File: src/lbaf/Utils/lbsJSONDataFilesMaker.py (hunk 11 above)
Comment (tlamonthezie) [reply src/lbaf/Utils/lbsJSONDataFilesMaker.py:241]: Fixed

File: src/lbaf/Utils/lbsJSONDataFilesMaker.py (hunk 10 above)
Comment (tlamonthezie) [reply src/lbaf/Utils/lbsJSONDataFilesMaker.py]: Fixed

File: src/lbaf/Utils/lbsJSONDataFilesMaker.py (hunk 12 above)
Comment (tlamonthezie) [reply src/lbaf/Utils/lbsJSONDataFilesMaker.py]: Fixed

**Comment by ppebay** [comment]
@nlslatt would you like to resolve the conversations, following @tlamonthezie 's changes? I can do it otherwise.

**Review by lifflander — APPROVED** [review]
Approving, looks good to me

**Comment by ppebay** [comment]
@tlamonthezie please feel free to merge ASAP, thanks

**Pull request merged**